compiler/src/iree/compiler/DispatchCreation/SetEncoding.cpp

```diff
@@ -306,10 +306,228 @@ struct FoldFillWithSetEncoding final
+  // TODO(MaheshRavishankar): Multi-result producer dispatches can be supported.
+  // Will require to move the consumer dispatch immediately after the producer
+  // instead of what is done below and move other operands of the consumer
+  // dispatch before the producer dispatch.
+  if (!producerDispatch ||
+      !llvm::hasSingleElement(producerDispatch.getBody()) ||
+      producerDispatch->getNumResults() != 1) {
+    return std::nullopt;
+  }
+  if (!llvm::hasSingleElement(producerValue.getUses())) {
+    return std::nullopt;
+  }
+
+  Location loc = producerDispatch->getLoc();
+  unsigned resultNumber = producerValue.getResultNumber();
+  // Compute the padding encoding
+  SmallVector<int64_t> paddingValue(operandType.getRank(), 0);
+  paddingValue.back() = ShapedType::kDynamic; // Set to dynamic for backend to
+                                              // pick the right value.
+  auto encoding = IREE::Encoding::PadEncodingLayoutAttr::get(
+      rewriter.getContext(), paddingValue);
+
+  // Compute the result types of the new dispatch.
+  auto newResultType = RankedTensorType::get(
+      operandType.getShape(), operandType.getElementType(), encoding);
+  auto newResultTypes = llvm::to_vector(producerDispatch->getResultTypes());
+  newResultTypes[resultNumber] = newResultType;
+
+  // Compute the result dynamic dims.
+  SmallVector<OpFoldResult> operandDims =
+      tensor::getMixedSizes(rewriter, loc, operand);
+  SmallVector<int64_t> operandStaticDims;
+  SmallVector<Value> operandDynamicDims;
+  dispatchIndexOpFoldResults(operandDims, operandDynamicDims,
+                             operandStaticDims);
+  SmallVector<Value> newResultDynamicDims;
+  for (OpResult producerDispatchResult : producerDispatch.getResults()) {
+    if (producerDispatchResult != producerValue) {
+      llvm::append_range(newResultDynamicDims,
+                         producerDispatch.getResultDynamicDims(
+                             producerDispatchResult.getResultNumber()));
+      continue;
+    }
+    llvm::append_range(newResultDynamicDims, operandDynamicDims);
+  }
+
+  auto newDispatchOp = rewriter.create<IREE::Flow::DispatchRegionOp>(
+      producerDispatch->getLoc(), newResultTypes, newResultDynamicDims,
+      producerDispatch.getWorkload());
+
+  // Move over the body of the old dispatch.
+  Region &newBody = newDispatchOp.getBody();
+  Region &producerDispatchBody = producerDispatch.getBody();
+  rewriter.cloneRegionBefore(producerDispatchBody, newBody, newBody.begin());
+
+  // Move over the slice operations if needed.
+  Region &producerWorkgroupCountBody = producerDispatch.getWorkgroupCount();
+  if (!producerWorkgroupCountBody.empty()) {
+    Region &newWorkgroupCountBody = newDispatchOp.getWorkgroupCount();
+    rewriter.cloneRegionBefore(producerWorkgroupCountBody,
+                               newWorkgroupCountBody,
+                               newWorkgroupCountBody.begin());
+  }
```

Comment:
Do we really need to handle the `WorkgroupCount` region? I think we usually don't have the region when we set encodings.

Reply:
I think we need it for completeness. If there was a region then dropping that would be a compiler error.

Reply:
Let me rephrase my question. Do we set encodings when there is a workgroup count region? Usually, the SetEncoding pass runs before workgroup count region creation. So my question is more like: does it mean that the fusion strategy is decided, if the workgroup count region is present? If the answer is no, the implementation looks good. If the answer is yes, we should not set encodings in this case.

Reply:
ok, I thought this might be what you had in mind. I think we should not set encoding in such cases (where workgroup count is already present), but I dont have an example to go by either way. It is a lose indication that workgroup count region being present means something, and I dont want to attach too much meaning to that. For now I am just doing the simplest "right" thing. We should revisit this later though.

Reply:
I see, thanks!